### ISSUE TITLE
Auth provider for Linkedin not working

### ISSUE THREAD

**Issue by andersengenolsen** [nextauthjs/next-auth#8831]
### Environment

System:
    OS: Linux 6.2 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
    CPU: (12) x64 AMD Ryzen 5 3600X 6-Core Processor
    Memory: 2.63 GB / 15.53 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 20.4.0 - /usr/local/bin/node
    npm: 9.7.2 - /usr/local/bin/npm

### Reproduction URL

https://github.com/nextauthjs/next-auth-example

### Describe the issue

As per the documentation from Linkedin, I've set up a new LinkedIn app, and added "Sign In with LinkedIn using OpenID Connect" as a product.

At first I had some problems when not specifying scope.

``
    providers: [
        LinkedIn({
            clientId: process.env.LINKEDIN_ID,
            clientSecret: process.env.LINKEDIN_SECRET,
        })
    ], pages: {
        signIn: '/register-cv'
    },
``

This returns an 'unauthorized_scope_error' for r_emailaddress. Managed to fix that issue by providing scopes as per the documentation from Microsoft:

> Authenticating Members
New members logging in to your service for the first time will need to follow the Authenticating with OAuth 2.0 Guide. When requesting the authorization code in Step 2 of the OAuth 2.0 Guide, make sure you use the OpenID scope openid to get the ID Token. We are also introducing new scopes profile and email.

> openid	Required to indicate the application wants to use OIDC to authenticate the member.
> profile	Required to retrieve the member's lite profile including their id, name, and profile picture.
> email	Required to retrieve the member's email address.
> After successful authentication, you will receive the member's access token and ID token.

> If your application does not have these permissions provisioned, you can request access through the Developer Portal. Select your app from My Apps, navigate to the Products tab, and request the Sign in with LinkedIn using OpenID Connect product.

Link: https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin-v2

Updated scopes to match the documentation from Microsoft:

`LinkedIn({
            clientId: process.env.LINKEDIN_ID,
            clientSecret: process.env.LINKEDIN_SECRET,
            authorization: { params: { scope: 'openid profile email' } }})`
            
Now getting this error:

`https://next-auth.js.org/errors#oauth_callback_error unexpected iss value, expected undefined, got: https://www.linkedin.com {
  error: RPError: unexpected iss value, expected undefined, got: https://www.linkedin.com
      at Client.validateJWT (/home/deb/PhpstormProjects/cvmaker/node_modules/openid-client/lib/client.js:931:15)
      at Client.validateIdToken (/home/deb/PhpstormProjects/cvmaker/node_modules/openid-client/lib/client.js:766:60)
      at Client.callback (/home/deb/PhpstormProjects/cvmaker/node_modules/openid-client/lib/client.js:505:18)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async oAuthCallback (/home/deb/PhpstormProjects/cvmaker/node_modules/next-auth/core/lib/oauth/callback.js:109:16)
      at async Object.callback (/home/deb/PhpstormProjects/cvmaker/node_modules/next-auth/core/routes/callback.js:52:11)
      at async AuthHandler (/home/deb/PhpstormProjects/cvmaker/node_modules/next-auth/core/index.js:208:28)
      at async NextAuthApiHandler (/home/deb/PhpstormProjects/cvmaker/node_modules/next-auth/next/index.js:22:19)
      at async NextAuth._args$ (/home/deb/PhpstormProjects/cvmaker/node_modules/next-auth/next/index.js:108:14) {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'linkedin',
  message: 'unexpected iss value, expected undefined, got: https://www.linkedin.com'
`

![image](https://github.com/nextauthjs/next-auth/assets/27057646/d113d3c3-d85c-4026-b843-5920579dfcba)

### How to reproduce

1: Set up new app on Linkedin, add "Sign In with LinkedIn using OpenID Connect" as a product.
2: Add authentication provider for Linkedin:

```import NextAuth from "next-auth"
import LinkedIn from "next-auth/providers/linkedin"

export const authOptions = {
    secret: 'NOT SO SECRET ANYMORE',
    providers: [
        LinkedIn({
            clientId: process.env.LINKEDIN_ID,
            clientSecret: process.env.LINKEDIN_SECRET,
            authorization: { params: { scope: 'openid profile email' } },
        })
    ], pages: {
        signIn: '/register-cv'
    },
}
export default `NextAuth(authOptions);

**Comment by andersengenolsen** [comment]
## SOLVED

After some debugging I managed to solve this. In case anyone else stumbles across this, since it's not well documented:

If you're using Linkedin API V2, you do need to add some custom parameters to the Linkedin provider. The original issue was that the expected iss was undefined. Simply add issuer to OAuthConfig for the Linkedin provider:

`issuer: 'https://www.linkedin.com'`

This triggers a new error: 'jwks_uri must be configured on the issuer'. Simply add jwks_endpoint to OAuthConfig:

`jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks'`

Again, this triggers a new error, where nextauth complains that the profile id is missing. Linkedin API V2 use "sub" not "id". Solved by overriding the profile function in OAuthConfig:

```async profile(profile) {
             return {
                    id: profile.sub,
                    name: profile.name,
                    firstname: profile.given_name,
                    lastname: profile.family_name,
                    email: profile.email,
                    image: profile.picture,
                }
            }
```

## Complete working provider for Linkedin using Linkedin API V2:

```
LinkedIn({
            clientId: process.env.LINKEDIN_ID,
            clientSecret: process.env.LINKEDIN_SECRET,
            authorization: { params: { scope: 'profile email openid' } },
            issuer: 'https://www.linkedin.com',
            jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
            async profile(profile) {
                return {
                    id: profile.sub,
                    name: profile.name,
                    firstname: profile.given_name,
                    lastname: profile.family_name,
                    email: profile.email
                }
            },
        })
```

**Comment by goerlitz** [comment]
This issue should be opened again, because every new app using SignIn with Linked will run into this issue (as I did).

Reason: **Microsoft has deprecated the old API as of August 1, 2023!**
- https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin
- https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin-v2

I'd suggest to keep the current `LinkedIn` provider as is (for all apps using the legacy API), but add a new `LinkedInV2` provider for all new apps using the _SignIn with OpenID Connect_.

## Deprecated API:
![LinkedIn_deprecated](https://github.com/nextauthjs/next-auth/assets/2389436/5147ff98-df1e-4909-b150-605606cf06f8)
![LinkedIn_deplrecated_scopes](https://github.com/nextauthjs/next-auth/assets/2389436/b64014b5-d16a-46ab-b1e4-7ef330f96d02)


## New API
![LinkedIn_new](https://github.com/nextauthjs/next-auth/assets/2389436/83a96c89-dc7f-4ab7-b310-8ca34a574f19)
![LinkedIn_new_scopes](https://github.com/nextauthjs/next-auth/assets/2389436/c3aa6e8f-f57e-44bb-9690-1a5cd1523800)

**Comment by andersengenolsen** [comment]
Hi,

Fair point. Closed since I managed to solve it by setting some custom parameters. although I used quite some time on it. Reopening.

Either way: Did you manage to get it to work using the provided solution?

**Comment by goerlitz** [comment]
Yes, your solution works great. It saved me a lot of time digging deeper myself. :)

**Comment by jormaj** [comment]
Is it still working for you guys? Even with the snippet posted, I'm still getting an error:
`error: invalid_client,
  error_description: Client authentication failed`

(I've double checked the client_id and client_secret multiple times already..

**Comment by jormaj** [comment]
Thanks for you reply :)
I got one step 'further': I was using auth.js version 0.12.0. After updating to 0.17.0, I'm now getting a different error:

```[mf:inf] GET /auth/csrf 200 OK (1ms)
Called .text() on an HTTP body which does not appear to be text. The body's Content-Type is "application/x-www-form-urlencoded". The result will probably be corrupted. Consider checking the Content-Type header before interpreting entities as text.
[auth][error][SignInError]: Read more at https://errors.authjs.dev#signinerror
[auth][cause]: OperationProcessingError: "response" is not a conform Authorization Server Metadata response
    at processDiscoveryResponse (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:5277:11)
    at getAuthorizationUrl (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:6903:23)
    at async signin (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:6966:14)
    at async AuthInternal (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:7144:27)
    at async Auth (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:7240:28)
    at async respond (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:17685:22)
    at async Object.fetch (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18172:13)
    at async jsonError (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18195:12)
    at async jsonError2 (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18373:12)
[auth][details]: {
  "provider": "linkedin"
}
[mf:inf] POST /auth/signin/linkedin 200 OK (133ms)
[mf:inf] GET /auth/signin 200 OK (2ms)
```


My code:  


```SvelteKitAuth(async (event) => {
    const authOptions = {
      providers: [GitHub({
        clientId: GITHUB_ID, clientSecret: GITHUB_SECRET
      }),
      Google({ clientId: GOOGLE_CLIENT_ID, clientSecret: GOOGLE_CLIENT_SECRET }),
      Facebook({ clientId: FACEBOOK_CLIENT_ID, clientSecret: FACEBOOK_CLIENT_SECRET }),
      LinkedIn({
        clientId: LINKEDIN_CLIENT_ID,
        clientSecret: LINKEDIN_CLIENT_SECRET,
        authorization: { params: { scope: 'profile email openid' } },
        issuer: 'https://www.linkedin.com',
        jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
        async profile(profile) {
          return {
            id: profile.sub,
            name: profile.name,
            firstname: profile.given_name,
            lastname: profile.family_name,
            email: profile.email
          }
        },
      })
      ],
      secret: AUTH_SECRET,
      trustHost: true,
      ...
      }
      ```

**Comment by goerlitz** [comment]
I'm using `"next-auth": "^4.23.2"` - not sure what that auth.js 0.17.0 dependency is.

**Comment by bholagabbar** [comment]
> ## SOLVED
> After some debugging I managed to solve this. In case anyone else stumbles across this, since it's not well documented:
> 
> If you're using Linkedin API V2, you do need to add some custom parameters to the Linkedin provider. The original issue was that the expected iss was undefined. Simply add issuer to OAuthConfig for the Linkedin provider:
> 
> `issuer: 'https://www.linkedin.com'`
> 
> This triggers a new error: 'jwks_uri must be configured on the issuer'. Add jwks_endpoint to OAuthConfig:
> 
> `jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks'`
> 
> Again, this triggers a new error, where nextauth complains that the profile id is missing. Linkedin API V2 use "sub" not "id". Solved by overriding the profile function in OAuthConfig:
> 
> ```
>              return {
>                     id: profile.sub,
>                     name: profile.name,
>                     firstname: profile.given_name,
>                     lastname: profile.family_name,
>                     email: profile.email,
>                     image: profile.picture,
>                 }
>             }
> ```
> 
> ## Complete working provider for Linkedin using Linkedin API V2:
> ```
> LinkedIn({
>             clientId: process.env.LINKEDIN_ID,
>             clientSecret: process.env.LINKEDIN_SECRET,
>             authorization: { params: { scope: 'profile email openid' } },
>             issuer: 'https://www.linkedin.com',
>             jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
>             async profile(profile) {
>                 return {
>                     id: profile.sub,
>                     name: profile.name,
>                     firstname: profile.given_name,
>                     lastname: profile.family_name,
>                     email: profile.email
>                 }
>             },
>         })
> ```

God bless, this is the right solution and docs should be updated

**Comment by mehul1011** [comment]
@andersengenolsen
If you have looked into the API deeply, what is the "sub" prop being returned by them in the response, and how can it be further used?
Also, there is any way to know the users' actual LinkedIn ID (for example: personx is the user ID below)?
https://linkedin.com/personx/

**Comment by andersengenolsen** [comment]
> @andersengenolsen If you have looked into the API deeply, what is the "sub" prop being returned by them in the response, and how can it be further used? Also, there is any way to know the users' actual LinkedIn ID (for example: personx is the user ID below)? https://linkedin.com/personx/

Hi,

The sub parameter is simply an unique user identifier issued "within" the ID token (JWT) from Linkedin API V2 after authenticating.

The latter seems to be a bit out of scope. Documentation available here: https://learn.microsoft.com/en-us/linkedin/consumer/

**Comment by mehul1011** [comment]
> > @andersengenolsen If you have looked into the API deeply, what is the "sub" prop being returned by them in the response, and how can it be further used? Also, there is any way to know the users' actual LinkedIn ID (for example: personx is the user ID below)? https://linkedin.com/personx/
> 
> Hi,
> 
> The sub parameter is simply an unique user identifier issued "within" the ID token (JWT) from Linkedin API V2 after authenticating.
> 
> The latter seems to be a bit out of scope. Documentation available here: https://learn.microsoft.com/en-us/linkedin/consumer/

Thanks for you response

Basically they have restricted most of the api to a normal developer i believe?
Like for example getting user profile information, also communication api

**Comment by GustavoContreiras** [comment]
I'm still receiving
```
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error read ECONNRESET {
  error: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20)
      at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    name: 'OAuthCallbackError',
    code: 'ECONNRESET'
  },
  providerId: 'linkedin',
  message: 'read ECONNRESET'
}
```
```
"next": "^13.5.6",
"next-auth": "^4.24.4",
```

**Comment by AnatoleT** [comment]
> Thanks for you reply :) I got one step 'further': I was using auth.js version 0.12.0. After updating to 0.17.0, I'm now getting a different error:
> 
> ```
> [mf:inf] GET /auth/csrf 200 OK (1ms)
> Called .text() on an HTTP body which does not appear to be text. The body's Content-Type is "application/x-www-form-urlencoded". The result will probably be corrupted. Consider checking the Content-Type header before interpreting entities as text.
> [auth][error][SignInError]: Read more at https://errors.authjs.dev#signinerror
> [auth][cause]: OperationProcessingError: "response" is not a conform Authorization Server Metadata response
>     at processDiscoveryResponse (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:5277:11)
>     at getAuthorizationUrl (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:6903:23)
>     at async signin (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:6966:14)
>     at async AuthInternal (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:7144:27)
>     at async Auth (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:7240:28)
>     at async respond (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:17685:22)
>     at async Object.fetch (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18172:13)
>     at async jsonError (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18195:12)
>     at async jsonError2 (file:///tmp/tmp-17164-QJe0X481iy9Q/hqlw55rde5g.js:18373:12)
> [auth][details]: {
>   "provider": "linkedin"
> }
> [mf:inf] POST /auth/signin/linkedin 200 OK (133ms)
> [mf:inf] GET /auth/signin 200 OK (2ms)
> ```
> 
> My code:
> 
> ```
> SvelteKitAuth(async (event) => {
>     const authOptions = {
>       providers: [GitHub({
>         clientId: GITHUB_ID, clientSecret: GITHUB_SECRET
>       }),
>       Google({ clientId: GOOGLE_CLIENT_ID, clientSecret: GOOGLE_CLIENT_SECRET }),
>       Facebook({ clientId: FACEBOOK_CLIENT_ID, clientSecret: FACEBOOK_CLIENT_SECRET }),
>       LinkedIn({
>         clientId: LINKEDIN_CLIENT_ID,
>         clientSecret: LINKEDIN_CLIENT_SECRET,
>         authorization: { params: { scope: 'profile email openid' } },
>         issuer: 'https://www.linkedin.com',
>         jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
>         async profile(profile) {
>           return {
>             id: profile.sub,
>             name: profile.name,
>             firstname: profile.given_name,
>             lastname: profile.family_name,
>             email: profile.email
>           }
>         },
>       })
>       ],
>       secret: AUTH_SECRET,
>       trustHost: true,
>       ...
>       }
>       ```
> ```

Hi @jormaj , did you find a workaround for this ? I'm facing the same error..

**Comment by AnatoleT** [comment]
> Is it still working for you guys? Even with the snippet posted, I'm still getting an error: `error: invalid_client, error_description: Client authentication failed`
> 
> (I've double checked the client_id and client_secret multiple times already..

Hi @jormaj , did you find a workaround for this ? I'm facing the same error..

**Comment by benjamindell** [comment]
Has anyone resolved this? I'm also getting the invalid_client error and can't work out why.

**Comment by AmitSahoo45** [comment]
Thanks. Resolved it. 👍🏻

**Comment by balazsorban44** [comment]
Hi everyone, I'm looking into this now. I think I found a bug in the LinkedIn OIDC implementation, I reached out to them and awaiting response.

**Comment by benjamindell** [comment]
> Hi everyone, I'm looking into this now. I think I found a bug in the LinkedIn OIDC implementation, I reached out to them and awaiting response.

Ah fantastic! Thanks for looking into this. Can confirm it's still an issue at my end.

**Comment by hexcowboy** [comment]
> Hi everyone, I'm looking into this now. I think I found a bug in the LinkedIn OIDC implementation, I reached out to them and awaiting response.

is this in reference to this error message?

> [auth][error] OperationProcessingError: "response" is not a conform Authorization Server Metadata response

because i'm also coming to the same conclusion

**Comment by lovinzoop** [comment]
> > ## SOLVED
> > After some debugging I managed to solve this. In case anyone else stumbles across this, since it's not well documented:
> > If you're using Linkedin API V2, you do need to add some custom parameters to the Linkedin provider. The original issue was that the expected iss was undefined. Simply add issuer to OAuthConfig for the Linkedin provider:
> > `issuer: 'https://www.linkedin.com'`
> > This triggers a new error: 'jwks_uri must be configured on the issuer'. Add jwks_endpoint to OAuthConfig:
> > `jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks'`
> > Again, this triggers a new error, where nextauth complains that the profile id is missing. Linkedin API V2 use "sub" not "id". Solved by overriding the profile function in OAuthConfig:
> > ```
> >              return {
> >                     id: profile.sub,
> >                     name: profile.name,
> >                     firstname: profile.given_name,
> >                     lastname: profile.family_name,
> >                     email: profile.email,
> >                     image: profile.picture,
> >                 }
> >             }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ## Complete working provider for Linkedin using Linkedin API V2:
> > ```
> > LinkedIn({
> >             clientId: process.env.LINKEDIN_ID,
> >             clientSecret: process.env.LINKEDIN_SECRET,
> >             authorization: { params: { scope: 'profile email openid' } },
> >             issuer: 'https://www.linkedin.com',
> >             jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
> >             async profile(profile) {
> >                 return {
> >                     id: profile.sub,
> >                     name: profile.name,
> >                     firstname: profile.given_name,
> >                     lastname: profile.family_name,
> >                     email: profile.email
> >                 }
> >             },
> >         })
> > ```
> 
> God bless, this is the right solution and docs should be updated

I tried this exact same configuration in LinkedInProvider in next-auth, but I am not getting user information.
![image](https://github.com/nextauthjs/next-auth/assets/123439986/3a52a995-9e66-4178-a938-9a37f54473bd)
![image](https://github.com/nextauthjs/next-auth/assets/123439986/f470a44e-6d24-460e-8f15-66450217f182)
In the signIn callback, console the response
![image](https://github.com/nextauthjs/next-auth/assets/123439986/93d16441-1111-46ed-8d49-824ff3c8ec5a)

**Comment by balazsorban44** [comment]
The main problem is that LinkedIn is currently not spec-compliant as [per the OIDC spec](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig).

`https://www.linkedin.com/oauth/.well-known/openid-configuration` returns `issuer` as `https://www.linkedin.com` while it should be `https://www.linkedin.com/oauth`, or the `https://www.linkedin.com/.well-known/openid-configuration` URL should not 404 but return the same as `https://www.linkedin.com/oauth/.well-known/openid-configuration` does. I reached out to them but haven't gotten a response yet.

While this is the case, I cannot easily update the default provider config. I'm thinking about adding a new `conform()` method to `issuer`, similarly to what `token` has: https://github.com/nextauthjs/next-auth/blob/26c47a8e1ae3e7b1069878003fe15da4e1a1928b/packages/core/src/providers/oauth.ts#L264C1-L265

**Comment by jacobwarren** [comment]
Is there any update for this @balazsorban44 ? In v5 beta 4 I'm still gettin the following error despite having implemented the fixes here:

```
[auth][error] OperationProcessingError: "response" is not a conform Authorization Server Metadata response
    at Module.processDiscoveryResponse (webpack-internal:///(rsc)/./node_modules/oauth4webapi/build/index.js:259:15)
    at getAuthorizationUrl (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/actions/signin/authorization-url.js:24:68)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.signIn (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/actions/signin/index.js:19:60)
    at async AuthInternal (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/index.js:71:24)
    at async Auth (webpack-internal:///(rsc)/./node_modules/@auth/core/index.js:123:29)
    at async eval (webpack-internal:///(rsc)/./node_modules/next/dist/esm/server/future/route-modules/app-route/module.js:218:37)
 ✓ Compiled in 491ms (2078 modules)
 ```
 
 Here's my config:
 
 ```
 import LinkedIn from "next-auth/providers/linkedin";

import type { NextAuthConfig } from "next-auth";

export default {
  providers: [
    LinkedIn({
      // clientId: process.env.LINKEDIN_ID!,
      // clientSecret: process.env.LINKEDIN_SECRET!,
      authorization: {
        params: {
          scope:
            "profile openid email w_member_social r_organization_social w_organization_social rw_organization_admin r_1st_connections_size",
        },
      },
      issuer: "https://www.linkedin.com",
      jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
      async profile(profile) {
        console.log("PROFILE FROM LINKEDIN:", profile);

        return {
          id: profile.sub,
          name: profile.name,
          firstname: profile.given_name,
          lastname: profile.family_name,
          email: profile.email,
        };
      },
    }),
  ],
} satisfies NextAuthConfig;
```

**Comment by balazsorban44** [comment]
No update, LinkedIn has not answered me yet. I pinged them again.

The solution for them would be a single rewrite from https://www.linkedin.com/.well-known/openid-configuration to https://www.linkedin.com/oauth/.well-known/openid-configuration to be spec-compliant.

**Comment by 100lvlmaster** [comment]
Is there a workaround for v5 beta 4 for now?

**Comment by 100lvlmaster** [comment]
Not even with a custom provider?

**Comment by balazsorban44** [comment]
We don't want to support non-spec OAuth providers, the industry needs to converge on specifications, especially if they are committing to it.

Of course you can hack around it, but if you are not sure how to, it's probably best to wait out. A PR for the proposed `conform()` method is welcome, if you are interested.

**Comment by jacobwarren** [comment]
Totally get this. At the same time, LinkedIn has proven they give zero cares about their API or conformity. Realistically, they'll never implement this properly. I'll take a stab at the `conform()` method in the next couple of weeks after I get a few things off my plate. Thanks for outlining the suggested fix, @balazsorban44 !

**Comment by balazsorban44** [comment]
I actually am in contact with them, I was just busy the latest reply was pending on my side. 🙈

**Comment by goerlitz** [comment]
@balazsorban44 will the fix work for LinkedIn's API V1 and V2 simultaneously?
I think that application still using the old V1 should not break whereas new apps have to use V2.

**Comment by balazsorban44** [comment]
Yeah, it would be possible to support both.

We did it with Twitter when they were transitioning from OAuth 1 to 2.0.

https://github.com/nextauthjs/next-auth/blob/2072750c017c3d2a7f7374dd683055b79d331319/packages/next-auth/src/providers/twitter.ts#L173

You had to pass `version: "2.0"`.

Here, I think you could just override `type: "oauth" | "oidc"` and we can switch the config based on that. (Ironically, I was [helping Twitter to ship OAuth 2.0](https://devcommunity.x.com/t/announcing-oauth-2-0-general-availability/163555) too 😄)

**Comment by martinratinaud** [comment]
I managed to make it work with the following (Note the `wellKnown`)

```
{
    ...LinkedInProvider({
      clientId: process.env.AUTH_LINKEDIN_CLIENT_ID as string,
      clientSecret: process.env.AUTH_LINKEDIN_CLIENT_SECRET as string,
    }),
    client: { token_endpoint_auth_method: 'client_secret_post' },
    issuer: 'https://www.linkedin.com',

    authorization: {
      url: 'https://www.linkedin.com/oauth/v2/authorization',
      params: { scope: 'openid profile email' },
    },
    wellKnown: 'https://www.linkedin.com/oauth/.well-known/openid-configuration',
    userinfo: {
      url: 'https://api.linkedin.com/v2/me',
      params: {
        projection: ``,
      },
    },
    async profile(profile: LinkedInProfile) {
      return {
        id: profile.sub,
        firstName: profile.given_name,
        lastName: profile.family_name,
        email: profile.email,
        image: profile.picture,
      };
    },
    allowDangerousEmailAccountLinking: true,
  };
```

Although, I tried to retrieve the `vanityName` by using the scope ` r_basicprofile` and found it impossible. Anybody managed to do so?

Thanks

**Comment by deibyrios** [comment]
@balazsorban44  I'm getting this error so I guess this is pending Linkedin to conform, right?: [auth][cause]: Error: TODO: Handle OIDC response body error

**Comment by alxwest** [comment]
@deibyrios Did you manage to get past this?

**Comment by deibyrios** [comment]
@alxwest not really. I’m not sure if @balazsorban44 is still working with LinkedIn for a fix. The other option option would be for someone to work on the conform() method but that’s beyond my current knowledge so, I guess we will have to dig into the issue and hack around Auth.JS to work, even with LinkedIn not being OIDC spec compliant.

**Comment by cocoBavan** [comment]
Putting logs inside node module files and patching things, I managed to get this far. 

`LinkedInProvider({
      clientId: process.env.LINKEDIN_CLIENT_ID,
      clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
      authorization: {
        url: "https://www.linkedin.com/oauth/v2/authorization",
        params: { scope: "profile email openid" },
      },
      token: {
        url: "https://www.linkedin.com/oauth/v2/accessToken",
      },
      userinfo: {
        url: "https://api.linkedin.com/v2/userinfo",
      },
    })`

But, it fails with the _**Error: TODO: Handle OIDC response body error**_. So, I believe the token returned is not correct and needs to fixed by Linked In?

**Comment by Shashwat61** [comment]
> Putting logs inside node module files and patching things, I managed to get this far.
> 
> ```
> LinkedInProvider({
>       clientId: process.env.LINKEDIN_CLIENT_ID,
>       clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
>       authorization: {
>         url: "https://www.linkedin.com/oauth/v2/authorization",
>         params: { scope: "profile email openid" },
>       },
>       token: {
>         url: "https://www.linkedin.com/oauth/v2/accessToken",
>       },
>       userinfo: {
>         url: "https://api.linkedin.com/v2/userinfo",
>       },
>     })
> ```
> 
> But, it fails with the _**Error: TODO: Handle OIDC response body error**_. So, I believe the token returned is not correct and needs to fixed by Linked In?

stuck at this point too

**Comment by paschalidi** [comment]
This has worked for me 
```
export const authOptions: AuthOptions = {
 .
 .
 providers: [
    LinkedInProvider({
      clientId: process.env.LINKEDIN_CLIENT_ID!,
      clientSecret: process.env.LINKEDIN_CLIENT_SECRET!,
      client: { token_endpoint_auth_method: 'client_secret_post' },
      authorization: {
        url: 'https://www.linkedin.com/oauth/v2/authorization',
        params: { scope: 'openid profile email' }
      },
      wellKnown:
        'https://www.linkedin.com/oauth/.well-known/openid-configuration',
      issuer: 'https://www.linkedin.com',
      jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks',
      async profile(profile: LinkedInProfile) {
        return {
          id: profile.sub,
          name: `${profile.given_name} ${profile.family_name}`,
          email: profile.email
        };
      }
    }),
  ]
 . 
 .
```

**Comment by paschalidi** [comment]
I only created it yesterday.

<img width="989" alt="image" src="https://github.com/nextauthjs/next-auth/assets/4181674/783dd26e-61f1-4810-add7-e35145107d47">

**Comment by cocoBavan** [comment]
@paschalidi Good for you. I even created a new one and tried it. Still the same error. Maybe the version of the next Auth or some other libraries has something to do with it?

**Comment by som-nitjsr** [comment]
check the latest code LinkedIn openid connect implementation. currently, there are not support the code_verfier parameter and claim dose not return noce.
by setting these two parameter shoud work.

    options.UsePkce = false;
    options.ProtocolValidator.RequireNonce = false;

**Comment by paramaggarwal** [comment]
@balazsorban44 I understand that auto discovery does not work as of today due to the mismatch in `issuer` and `wellKnown` from LinkedIn's side. But we should be able to bypass discovery and manually configure the endpoints in the default LinkedIn config of the library.


```
authorization: {
  url: "https://www.linkedin.com/oauth/v2/authorization",
  params: { scope: "openid profile email" },
},
token: {
  url: "https://www.linkedin.com/oauth/v2/accessToken",
},
userinfo: {
  url: "https://api.linkedin.com/v2/userinfo",
},
```

This should have worked. But @cocoBavan and @Shashwat61 and myself are all stuck at this stage with error:
> Error: TODO: Handle OIDC response body error

**Comment by balazsmeszegeto** [comment]
> check the latest code LinkedIn openid connect implementation. currently, there are not support the code_verfier parameter and claim dose not return noce. by setting these two parameter shoud work.
> 
> ```
> options.UsePkce = false;
> options.ProtocolValidator.RequireNonce = false;
> ```

@som-nitjsr Where did you get this from? It is correct, just asking for the source

**Comment by som-nitjsr** [comment]
> > check the latest code LinkedIn openid connect implementation. currently, there are not support the code_verfier parameter and claim dose not return noce. by setting these two parameter shoud work.
> > ```
> > options.UsePkce = false;
> > options.ProtocolValidator.RequireNonce = false;
> > ```
> 
> @som-nitjsr Where did you get this from? It is correct, just asking for the source

I have used c# .net core  and this is how i have integrated.

let me know which framework you are uisng 
 .AddOpenIdConnect("LinkedIn", "LinkedIn", options =>
 {
 

     options.Authority = "https://www.linkedin.com/oauth/";
     options.CallbackPath = "/signin-linkedin";
     options.ClientId = "clinetid";
     options.ClientSecret = "code";
     options.ResponseType = "code";
     options.Scope.Add("openid");
     options.Scope.Add("profile");
     options.Scope.Add("email");
     options.ProtocolValidator.RequireNonce = false;
     options.Events = new OpenIdConnectEvents
     {
         OnAuthorizationCodeReceived = async context =>
         {
             // Remove the 'code_verifier' parameter
             context.TokenEndpointRequest.Parameters.Remove("code_verifier");
             await Task.CompletedTask;
         }
     };
   
 });

**Comment by som-nitjsr** [comment]
also you can refer the c# code here where linkedin is connected uisng idp https://github.com/som-nitjsr/linkedidp

**Comment by balazsmeszegeto** [comment]
I am using ASP.NET Core, but couldn't figure out why PKCE would result in `invalid_client` error on token endpoint. Searching for the solution is how I ended up here and wanted to confirm, that it does indeed solve the issue.

> check the latest code LinkedIn openid connect implementation. currently, there are not support the code_verfier parameter and claim dose not return noce. by setting these two parameter shoud work.
> ```
> options.UsePkce = false;
> options.ProtocolValidator.RequireNonce = false;
> ```

Furthermore, when you said: "check the latest code LinkedIn openid connect implementation" - I had the impression is that you're referring to an official or other publicly available  implementation, so that's why I was asking for the source (the origin). Do you have that or did you figure this out on your own?

**Comment by som-nitjsr** [comment]
@balazsmeszegeto 
 i tested the token endpoint  by  passing the different parameters and figured out that code_verifer is not supported.

you can also see that they dont support nonce claim from here https://www.linkedin.com/oauth/.well-known/openid-configuration?_l=en_US

based on these i have written a solution here https://github.com/som-nitjsr/linkedidp

**Comment by balazsmeszegeto** [comment]
Great job then!

Still, I'd consider this as a bug in LinkedIn side, since nonce is mandatory as per OpenID C standard, and not an optional claim

**Comment by pranavmappoli** [comment]
Anyone find the final solution, still facing sam error

**Comment by Shashwat61** [comment]
> Anyone find the final solution, still facing same error

Just try using the strategy for LinkedIn as "oauth_linkedin_oidc", worked for me

**Comment by kafiln** [comment]
Anyone with a workaround for this ?

**Comment by AbdelrahmanSadat** [comment]
In case anyone comes across it, I had the same workaround in my codebase, until a different but similar error started showing `https://next-auth.js.org/errors#oauth_callback_error unexpected iss value, expected https://www.linkedin.com, got: https://www.linkedin.com/oauth`

I set `issuer: https://www.linkedin.com/oauth` in the `LinkedInProvider` and the error is resolved.

**Comment by mfts** [comment]
> I set issuer: 'https://www.linkedin.com/oauth' in the LinkedInProvider instead of issuer: 'https://www.linkedin.com',  and the error is resolved.

I can confirm this also worked for me. Must be a recent change in the API.

**Comment by balazsorban44** [comment]
Oh, it certainly is, I haven't gotten back a confirmation from them, but been asking for this for a while. Will fix it soon, now!

**Comment by ndom91** [comment]
Put up a PR with the change: https://github.com/nextauthjs/next-auth/pull/10671

**Comment by wowtah** [comment]
Did anybody got LinkedIn working for NextAuth.js v5? 

With my current code, I get the LinkedIn login page, and a LinkedIn page where I press 'Allow' (this also shows it will redirect to localhost). After choosing Allow, my browser (now on http://localhost:3000/api/auth/error?error=Configuration) shows:

_Server error
There is a problem with the server configuration.
Check the server logs for more information._

My logging shows:

```
GET /api/auth/signin 200 in 9ms
 POST /api/auth/signin/linkedin 302 in 32ms
error {
  error: 'invalid_client',
  error_description: 'Client authentication failed'
}
[auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle OIDC response body error
    at handleOAuth (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/actions/callback/oauth/callback.js:86:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.callback (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/actions/callback/index.js:32:41)
    at async AuthInternal (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/index.js:39:24)
    at async Auth (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/index.js:126:34)
    at async C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:53191
    at async e_.execute (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:44492)
    at async e_.handle (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:54445)
    at async doRender (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1377:42)
    at async cacheEntry.responseCache.get.routeKind (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1587:40)
    at async DevServer.renderToResponseWithComponentsImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1507:28)
    at async DevServer.renderPageComponent (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1924:24)
    at async DevServer.renderToResponseImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1962:32)
    at async DevServer.pipeImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:920:25)
    at async NextNodeServer.handleCatchallRenderRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\next-server.js:272:17)
    at async DevServer.handleRequestImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:816:17)
    at async C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\dev\next-dev-server.js:339:20
    at async Span.traceAsyncFn (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\trace\trace.js:154:20)
    at async DevServer.handleRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\dev\next-dev-server.js:336:24)
    at async invokeRender (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:174:21)
    at async handleRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:353:24)
    at async requestHandlerImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:377:13)
    at async Server.requestListener (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\start-server.js:141:13)
[auth][details]: {
  "provider": "linkedin"
}
 GET /api/auth/callback/linkedin?code=AQQzoxG4FredactedzW8aWageredactedYYmUsHXW3FsgQ9NbredactedwbF8D-TNbs_rdIbCbredactednZ_Dgy2RCZUBUYynredacteddICqBDJNBJhqUtFDDUWCEopZ-oypAH4gni5gN_PRnzmbh1OGgJP8NoxnUnhcAltC0oYz4_f5nA5ipY 302 in 319ms
 GET /api/auth/error?error=Configuration 500 in 12ms
```

I tried many variants of this code, some resulting in different errors. But still havent seen it working. 

```
LinkedInProvider({
        clientId: process.env.LINKEDIN_CLIENT_ID!,
        clientSecret: process.env.LINKEDIN_CLIENT_SECRET!,
        client: { token_endpoint_auth_method: 'client_secret_post',  },
        authorization: {
          url: 'https://www.linkedin.com/oauth/v2/authorization',
          params: { scope: 'openid profile email' }
        },
        token: {
          url: 'https://www.linkedin.com/oauth/v2/accessToken',
        },
        wellKnown:
          'https://www.linkedin.com/oauth/.well-known/openid-configuration',
          userinfo: {
            url: 'https://api.linkedin.com/v2/userinfo',
            params: {
              projection: ``,
            },
          },
        issuer: 'https://www.linkedin.com/oauth',
        jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks',
        //wellKnown: 'https://www.linkedin.com/oauth/.well-known/openid-configuration',
        
        profile(profile: LinkedInProfile) {
          return {
            id: profile.sub,
            firstName: profile.given_name,
            lastName: profile.family_name,
            email: profile.email,
            image: profile.picture,
          };
        },
        allowDangerousEmailAccountLinking: true
     
    })
```
I got it working for GitHub, so I am confident that my base setup is correct. 
The versions I am using:

```
"@auth/core": "^0.30.0",
"@auth/prisma-adapter": "^2.0.0",
"next": "^14.2.2",
"next-auth": "^5.0.0-beta.16",

```

I am posting this here since this thread seems very relevant. Any help is appreciated :)

**Comment by kafiln** [comment]
> Did anybody got LinkedIn working for NextAuth.js v5?
> 
> With my current code, I get the LinkedIn login page, and a LinkedIn page where I press 'Allow' (this also shows it will redirect to localhost). After choosing Allow, my browser (now on http://localhost:3000/api/auth/error?error=Configuration) shows:
> 
> _Server error There is a problem with the server configuration. Check the server logs for more information._
> 
> My logging shows:
> 
> ```
> GET /api/auth/signin 200 in 9ms
>  POST /api/auth/signin/linkedin 302 in 32ms
> error {
>   error: 'invalid_client',
>   error_description: 'Client authentication failed'
> }
> [auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
> [auth][cause]: Error: TODO: Handle OIDC response body error
>     at handleOAuth (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/actions/callback/oauth/callback.js:86:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Module.callback (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/actions/callback/index.js:32:41)
>     at async AuthInternal (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/lib/index.js:39:24)
>     at async Auth (webpack-internal:///(rsc)/./node_modules/next-auth/node_modules/@auth/core/index.js:126:34)
>     at async C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:53191
>     at async e_.execute (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:44492)
>     at async e_.handle (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:54445)
>     at async doRender (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1377:42)
>     at async cacheEntry.responseCache.get.routeKind (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1587:40)
>     at async DevServer.renderToResponseWithComponentsImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1507:28)
>     at async DevServer.renderPageComponent (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1924:24)
>     at async DevServer.renderToResponseImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:1962:32)
>     at async DevServer.pipeImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:920:25)
>     at async NextNodeServer.handleCatchallRenderRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\next-server.js:272:17)
>     at async DevServer.handleRequestImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\base-server.js:816:17)
>     at async C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\dev\next-dev-server.js:339:20
>     at async Span.traceAsyncFn (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\trace\trace.js:154:20)
>     at async DevServer.handleRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\dev\next-dev-server.js:336:24)
>     at async invokeRender (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:174:21)
>     at async handleRequest (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:353:24)
>     at async requestHandlerImpl (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\router-server.js:377:13)
>     at async Server.requestListener (C:\Users\Wouter\source\repos\nxt14\node_modules\next\dist\server\lib\start-server.js:141:13)
> [auth][details]: {
>   "provider": "linkedin"
> }
>  GET /api/auth/callback/linkedin?code=AQQzoxG4FredactedzW8aWageredactedYYmUsHXW3FsgQ9NbredactedwbF8D-TNbs_rdIbCbredactednZ_Dgy2RCZUBUYynredacteddICqBDJNBJhqUtFDDUWCEopZ-oypAH4gni5gN_PRnzmbh1OGgJP8NoxnUnhcAltC0oYz4_f5nA5ipY 302 in 319ms
>  GET /api/auth/error?error=Configuration 500 in 12ms
> ```
> 
> I tried many variants of this code, some resulting in different errors. But still havent seen it working.
> 
> ```
> LinkedInProvider({
>         clientId: process.env.LINKEDIN_CLIENT_ID!,
>         clientSecret: process.env.LINKEDIN_CLIENT_SECRET!,
>         client: { token_endpoint_auth_method: 'client_secret_post',  },
>         authorization: {
>           url: 'https://www.linkedin.com/oauth/v2/authorization',
>           params: { scope: 'openid profile email' }
>         },
>         token: {
>           url: 'https://www.linkedin.com/oauth/v2/accessToken',
>         },
>         wellKnown:
>           'https://www.linkedin.com/oauth/.well-known/openid-configuration',
>           userinfo: {
>             url: 'https://api.linkedin.com/v2/userinfo',
>             params: {
>               projection: ``,
>             },
>           },
>         issuer: 'https://www.linkedin.com/oauth',
>         jwks_endpoint: 'https://www.linkedin.com/oauth/openid/jwks',
>                 
>         profile(profile: LinkedInProfile) {
>           return {
>             id: profile.sub,
>             firstName: profile.given_name,
>             lastName: profile.family_name,
>             email: profile.email,
>             image: profile.picture,
>           };
>         },
>         allowDangerousEmailAccountLinking: true
>      
>     })
> ```
> 
> I got it working for GitHub, so I am confident that my base setup is correct. The versions I am using:
> 
> ```
> "@auth/core": "^0.30.0",
> "@auth/prisma-adapter": "^2.0.0",
> "next": "^14.2.2",
> "next-auth": "^5.0.0-beta.16",
> ```
> 
> I am posting this here since this thread seems very relevant. Any help is appreciated :)

I am in the exact same situation.

**Comment by Starefossen** [comment]
Same configuration as @kafiln and @wowtah with the following error after callback from LinkedIn:

```
 POST /api/auth/signin/linkedin 302 in 24ms                                                                                       error {                                                          
  error: 'invalid_client',                                                                                                          error_description: 'Client authentication failed'
}                                                                                                                                 [auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle OIDC response body error 
```

Config:

```
export const config = {
  providers: [
    GitHub({
      clientId: process.env.AUTH_GITHUB_ID,
      clientSecret: process.env.AUTH_GITHUB_SECRET,
      authorization: { params: { scope: 'openid profile email' } },
    }),
    LinkedIn({
      clientId: process.env.AUTH_LINKEDIN_ID!,
      clientSecret: process.env.AUTH_LINKEDIN_SECRET!,
      authorization: {
        url: 'https://www.linkedin.com/oauth/v2/authorization',
        params: { scope: 'openid profile email' }
      },
      issuer: "https://www.linkedin.com/oauth",
   })
  ]
}
```

**Comment by Ranjanlc** [comment]
Still facing this issue. Am I doing something wrong? As there are lots of solutions in this thread, still unsure what is the correct one. This is my LinkedIn provider and I have latest version of next-auth 4.24.7 

And this is my linkedIn provider config:

```javascript
    LinkedInProvider({
        clientId: process.env.LINKEDIN_CLIENT_ID || "",
        clientSecret: process.env.LINKEDIN_CLIENT_SECRET || "",
        client: { token_endpoint_auth_method: "client_secret_post" },
        authorization: {
          url: "https://www.linkedin.com/oauth/v2/authorization",
          params: { scope: "openid profile email" },
        },
        token: {
          url: "https://www.linkedin.com/oauth/v2/accessToken",
        },
        userinfo: {
          url: "https://api.linkedin.com/v2/userinfo",
        },
        wellKnown:
          "https://www.linkedin.com/oauth/.well-known/openid-configuration",
        issuer: "https://www.linkedin.com/oauth",
        jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
        profile(profile) {
          const defaultImage =
            "https://cdn-icons-png.flaticon.com/512/174/174857.png";
          return {
            id: profile.sub,
            name: profile.name,
            email: profile.email,
            image: profile.picture ?? defaultImage,
          };
        },
      }),
```

**Comment by holgerflick** [comment]
I am also still getting an error:

```
[auth][error] OperationProcessingError: "response" is not a conform Authorization Server Metadata response
    at Module.processDiscoveryResponse (webpack-internal:///(rsc)/./node_modules/oauth4webapi/build/index.js:287:15)
    at getAuthorizationUrl (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/actions/signin/authorization-url.js:25:68)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.signIn (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/actions/signin/index.js:16:56)
    at async AuthInternal (webpack-internal:///(rsc)/./node_modules/@auth/core/lib/index.js:72:24)
    at async Auth (webpack-internal:///(rsc)/./node_modules/@auth/core/index.js:126:34)
    at async /Users/holger/web/next.providers/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:53446
    at async e_.execute (/Users/holger/web/next.providers/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:44747)
    at async e_.handle (/Users/holger/web/next.providers/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:54700)
    at async doRender (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:1377:42)
    at async cacheEntry.responseCache.get.routeKind (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:1587:40)
    at async DevServer.renderToResponseWithComponentsImpl (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:1507:28)
    at async DevServer.renderPageComponent (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:1924:24)
    at async DevServer.renderToResponseImpl (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:1962:32)
    at async DevServer.pipeImpl (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:920:25)
    at async NextNodeServer.handleCatchallRenderRequest (/Users/holger/web/next.providers/node_modules/next/dist/server/next-server.js:272:17)
    at async DevServer.handleRequestImpl (/Users/holger/web/next.providers/node_modules/next/dist/server/base-server.js:816:17)
    at async /Users/holger/web/next.providers/node_modules/next/dist/server/dev/next-dev-server.js:339:20
    at async Span.traceAsyncFn (/Users/holger/web/next.providers/node_modules/next/dist/trace/trace.js:154:20)
    at async DevServer.handleRequest (/Users/holger/web/next.providers/node_modules/next/dist/server/dev/next-dev-server.js:336:24)
    at async invokeRender (/Users/holger/web/next.providers/node_modules/next/dist/server/lib/router-server.js:174:21)
    at async handleRequest (/Users/holger/web/next.providers/node_modules/next/dist/server/lib/router-server.js:353:24)
    at async requestHandlerImpl (/Users/holger/web/next.providers/node_modules/next/dist/server/lib/router-server.js:377:13)
    at async Server.requestListener (/Users/holger/web/next.providers/node_modules/next/dist/server/lib/start-server.js:141:13)
 POST /api/auth/signin/linkedin 302 in 172ms
 GET /api/auth/error?error=Configuration 500 in 5ms
```

**Comment by Starefossen** [comment]
`npm i --save next-auth@^5.0.0-beta.18` did the trick for me! Thank you @balazsorban44 ❤️

**Comment by wowtah** [comment]
Fixed for me also in 5-beta.18. Thanks!

**Comment by thelastbackspace** [comment]
> In case anyone comes across it, I had the same workaround in my codebase, until a different but similar error started showing: `https://next-auth.js.org/errors#oauth_callback_error unexpected iss value, expected https://www.linkedin.com, got: https://www.linkedin.com/oauth`
> 
> I set `issuer: 'https://www.linkedin.com/oauth'` in the LinkedInProvider instead of `issuer: 'https://www.linkedin.com', ` and the error is resolved.

This is the solution that actually works

**Comment by seldo** [comment]
This combines all the feedback I saw above and is what works for me as of August 2024:

```
const handler = NextAuth({
    providers: [
        LinkedInProvider({
            clientId: process.env.LINKEDIN_CLIENT_ID,
            clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
            authorization: { params: { scope: 'profile email openid' } },
            issuer: 'https://www.linkedin.com/oauth',
            jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
            async profile(profile) {
                return {
                    id: profile.sub,
                    name: profile.name,
                    firstname: profile.given_name,
                    lastname: profile.family_name,
                    email: profile.email
                }
            },
        }),
    ],
});
```

**Comment by amahony** [comment]
> This combines all the feedback I saw above and is what works for me as of August 2024:
> 
> ```
> const handler = NextAuth({
>     providers: [
>         LinkedInProvider({
>             clientId: process.env.LINKEDIN_CLIENT_ID,
>             clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
>             authorization: { params: { scope: 'profile email openid' } },
>             issuer: 'https://www.linkedin.com/oauth',
>             jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
>             async profile(profile) {
>                 return {
>                     id: profile.sub,
>                     name: profile.name,
>                     firstname: profile.given_name,
>                     lastname: profile.family_name,
>                     email: profile.email
>                 }
>             },
>         }),
>     ],
> });
> ```

this worked for me perfectly!

**Comment by mmgolfdistrict** [comment]
After Adding All configuration of linkedin 
  LinkedInProvider({
            clientId: process.env.LINKEDIN_CLIENT_ID,
            clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
            authorization: { params: { scope: 'profile email openid' } },
            issuer: 'https://www.linkedin.com/oauth',
            jwks_endpoint: "https://www.linkedin.com/oauth/openid/jwks",
            async profile(profile) {
                return {
                    id: profile.sub,
                    name: profile.name,
                    firstname: profile.given_name,
                    lastname: profile.family_name,
                    email: profile.email
                }
            },
        }),
 above im still getting error :- 
error {
   error: 'invalid_client',
  error_description: 'Client authentication failed'
}
1: [auth][error][CallbackRouteError]: Read more at https://errors.authjs.dev#callbackrouteerror
 [auth][cause]: Error: TODO: Handle OIDC response body error
    at handleOAuth (webpack-internal:///(rsc)/../../node_modules/@auth/core/lib/oauth/callback.js:80:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.callback (webpack-internal:///(rsc)/../../node_modules/@auth/core/lib/routes/callback.js:34:41)
  at async AuthInternal (webpack-internal:///(rsc)/../../node_modules/@auth/core/lib/index.js:104:38)
   at async Auth (webpack-internal:///(rsc)/../../node_modules/@auth/core/index.js:121:30)
    at async /Users/manavgd/Documents/all-inclusive-mono-copy/all-inclusive-mono/apps/web/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:63251
[auth][details]: {
 "provider": "linkedin"

**Comment by caposada** [comment]
Yeah, thanks for the solution.

This worked for me as well with this config:

...
LinkedInProvider({
    clientId: process.env.AUTH_LINKEDIN_ID ?? "",
    clientSecret: process.env.AUTH_LINKEDIN_SECRET ?? "",
    issuer: "https://www.linkedin.com",
    profile: (profile: LinkedInProfile) => ({
      id: profile.sub,
      name: profile.name,
      email: profile.email,
      image: profile.picture,
    }),
    wellKnown:
      "https://www.linkedin.com/oauth/.well-known/openid-configuration",
    authorization: {
      params: {
        scope: "openid profile email",
      },
    },
  }),
...

Using NextAuth version 4.24.11

**Comment by YuriDev-creator** [comment]
> Yeah, thanks for the solution.
> 
> This worked for me as well with this config:
> 
> ... LinkedInProvider({ clientId: process.env.AUTH_LINKEDIN_ID ?? "", clientSecret: process.env.AUTH_LINKEDIN_SECRET ?? "", issuer: "https://www.linkedin.com", profile: (profile: LinkedInProfile) => ({ id: profile.sub, name: profile.name, email: profile.email, image: profile.picture, }), wellKnown: "https://www.linkedin.com/oauth/.well-known/openid-configuration", authorization: { params: { scope: "openid profile email", }, }, }), ...
> 
> Using NextAuth version 4.24.11

Funcionou aqui também nextAuth 4.24.11  -- 2025

**Comment by jrpursuiteam** [comment]
Hi, this is still not working for me. Any further lead?
